### PR TITLE
Use `ResultStore` by default for task transactions

### PR DESCRIPTION
**Pull request by desertaxle**
This PR updates the `Transaction` used for a task run to use a `ResultStore` instead of a `RecordStore`. Once a top-level method of `ResultStore` configuration is exposed, users will be able to configure result persistence, metadata persistence, and locking for task runs by changing the storage blocks and lock manager given to a `ResultStore`.

As a result of these changes, a JSON representation of `ResultMetadata` is stored in the `data` column of the `task_run_state` table. This replaces the JSON representation of `PersistedResult` that was used before. This maintains the current functionality to retrieve a persisted result from a state. The existing mechanics are kept for backwards compatibility.

A couple of key API additions/changes to make this possible:
- `should_persist_result`
    - The context for if results should be persisted or not is moved to the `FlowRunContext` and `TaskRunContext` models. This removes the responsibility of whether or not to persist from the `ResultStore`. Consumers of a `ResultStore` need to choose if `.write` or `.persist_result_record` should be called, and `should_persist_result` helps with that. The `persist_result` attribute is kept on `ResultStore` for backwards compatibility, but it has been marked as deprecated.
- `ResultRecord.from_metadata`
    - This method is the equivalent of `PersistedResult.get` and is used to fetch persisted results from their persisted location.
- `Transaction.write_on_commit`
    - This attribute controls whether a transaction should write a result record based on its staged value on commit. Since `Transaction` is a consumer of `ResultStore`, it is responsible for choosing when to delegate writes to the `ResultStore`.

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/client/schemas/objects.py [hunk 1]
```diff
@@ -168,9 +179,14 @@ class State(ObjectBaseModel, Generic[R]):
...
-    data: Union["BaseResult[R]", Any] = Field(
-        default=None,
-    )
+    data: Annotated[
+        Union[
+            Annotated["BaseResult[R]", Tag("BaseResult")],
+            Annotated["ResultRecordMetadata", Tag("ResultRecordMetadata")],
+            Annotated[Any, Tag("Any")],
+        ],
+        Discriminator(data_discriminator),
```
Comment (chrisguidry) [inline src/prefect/client/schemas/objects.py:188]: Tasty!

**Comment by desertaxle** [comment]
> One question: are there any supported cases were an older client could be trying to deserialize a persisted result from a newer client? I'm thinking like cases where a Prefect 2.20 client is submitting background tasks to a Prefect 3 task worker and resolving futures, that kind of thing. Is there anything we could add to 2.x/3.0 to give an appropriate warning/error when accessing a result that was saved with the new result store?

Good question! I think a scenario like that is unlikely since most result retrieval happens within the context of the same version, and background tasks are experimental in the 2.x line. The formats between new and old are very similar, though, and we could add some compatibility code to 2.x to gracefully handle the new format. We already have compatibility for old formats in the 3.x line.

File: src/prefect/flow_engine.py [hunk 2]
```diff
@@ -507,10 +512,13 @@ def setup_run_context(self, client: Optional[SyncPrefectClient] = None):
...
                     parameters=self.parameters,
                     client=client,
-                    result_store=get_current_result_store().update_for_flow(
+                    result_store=get_result_store().update_for_flow(
                         self.flow, _sync=True
                     ),
                     task_runner=task_runner,
+                    persist_result=self.flow.persist_result
+                    if self.flow.persist_result is not None
+                    else should_persist_result(),
```
Comment (cicdw) [inline src/prefect/flow_engine.py:521]: this is really clean, it ensures good inheritance in all nested scenarios too

File: src/prefect/records/filesystem.py [hunk 3]
```diff
@@ -70,7 +69,6 @@ def _get_lock_info(self, key: str, use_cache=True) -> Optional[_LockInfo]:
                     pendulum.parse(expiration) if expiration is not None else None
                 )
             self._locks[key] = lock_info
-            print("Got lock info from file")
```
Comment (cicdw) [inline src/prefect/records/filesystem.py:73]: my code relies on this print, this is a breaking change for me. 

jk jk

File: src/prefect/results.py [hunk 4]
```diff
@@ -171,13 +178,42 @@ def get_default_persist_setting() -> bool:
     return PREFECT_RESULTS_PERSIST_BY_DEFAULT.value()
 
 
+def should_persist_result() -> bool:
```
Comment (cicdw) [inline src/prefect/results.py:181]: i really like this utility lol

**Review by cicdw — APPROVED** [review]
this makes sense to me!

**Pull request merged**